crypto/kyber/p_kyber_test.cc

```diff
@@ -4,280 +4,463 @@
+    EVP_PKEY_CTX *kyber_pkey_ctx_alice = EVP_PKEY_CTX_new_id(pkey_id, nullptr);
+    EVP_PKEY *kyber_pkey_alice = EVP_PKEY_new();
+    EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx_alice));
+    EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx_alice, &kyber_pkey_alice));
+
+    // Alice passes the public key to Bob.
+    uint8_t *pub_alice = get_pub_from_kyber_key(kyber_pkey_alice);
+    EVP_PKEY_CTX *kyber_pkey_ctx_bob = EVP_PKEY_CTX_new_id(pkey_id, nullptr);
+    EVP_PKEY *kyber_pkey_bob = EVP_PKEY_new_raw_public_key(pkey_id, NULL, pub_alice, public_key_len);
+    kyber_pkey_ctx_bob->pkey = kyber_pkey_bob;
+
+    // Bob generates a shared secret and encapsulates it.
+    ASSERT_TRUE(EVP_PKEY_encapsulate(kyber_pkey_ctx_bob, ciphertext_bob, &ciphertext_len, shared_secret_bob, &shared_secret_len));
+
+    // Bob sends the ciphertext to Alice.
+    OPENSSL_memcpy(ciphertext_alice, ciphertext_bob, ciphertext_len);
+
+    // Alice decapsulates the ciphertext to obtain the shared secret.
+    ASSERT_TRUE(EVP_PKEY_decapsulate(kyber_pkey_ctx_alice, shared_secret_alice, &shared_secret_len, ciphertext_alice, ciphertext_len));
+
+    // Verify that Alice and Bob have the same shared secret.
+    for (size_t i = 0; i < shared_secret_len; i++) {
+      EXPECT_EQ(shared_secret_alice[i], shared_secret_bob[i]);
+    }
+
+    // Invalidate the ciphertext and run decapsulate on it.
+    ciphertext_alice[0] ^= 1;
+
+    // Decapsulate should return success.
+    ASSERT_TRUE(EVP_PKEY_decapsulate(kyber_pkey_ctx_alice, shared_secret_alice, &shared_secret_len, ciphertext_alice, ciphertext_len));
+
+    // But the shared secret should be different from Bob's.
+    unsigned char tmp = 0;
+    for (size_t i = 0; i < shared_secret_len; i++) {
+      tmp |= (shared_secret_alice[i] ^ shared_secret_bob[i]);
+    }
+    EXPECT_NE(tmp, 0);
+
+    OPENSSL_free(shared_secret_alice);
+    OPENSSL_free(shared_secret_bob);
+    OPENSSL_free(ciphertext_alice);
+    OPENSSL_free(ciphertext_bob);
+
+    EVP_PKEY_CTX_free(kyber_pkey_ctx_alice);
+    EVP_PKEY_CTX_free(kyber_pkey_ctx_bob);
+  }
+}
 
-  // Decapsulate should return success.
-  ASSERT_TRUE(EVP_PKEY_decapsulate(kyber_pkey_ctx_alice, shared_secret_alice, &shared_secret_len, ciphertext_alice, ciphertext_len));
+TEST(KyberTest, KEMSizeChecks) {
+  for (auto pkey_id : pkeyIDsToTest) {
+    size_t shared_secret_len = 0;
+    size_t ciphertext_len = 0;
+
+    size_t expected_ss_len = kyber_shared_secret_size(pkey_id);
+    size_t expected_ct_len = kyber_ciphertext_size(pkey_id);
+
+    EVP_PKEY_CTX *kyber_pkey_ctx = EVP_PKEY_CTX_new_id(pkey_id, nullptr);
+    EVP_PKEY *kyber_pkey = EVP_PKEY_new();
+    EXPECT_TRUE(EVP_PKEY_keygen_init(kyber_pkey_ctx));
+    EXPECT_TRUE(EVP_PKEY_keygen(kyber_pkey_ctx, &kyber_pkey));
+
+    ASSERT_TRUE(EVP_PKEY_encapsulate(kyber_pkey_ctx, NULL, &ciphertext_len, NULL, &shared_secret_len));
+    EXPECT_EQ(shared_secret_len, expected_ss_len);
+    EXPECT_EQ(ciphertext_len, expected_ct_len);
```

Comment:
Ahhh I see this test needs a second source of truth to compare EVP_PKEY_encapsulate with null pointers returns the expected size. Do you think it's worth having this test but need to create a second kem -> size mapping up in `kyber_ciphertext_size`?

Maybe if we go the full parametrized route it gets encoded into the test like the `KnownAEAD` struct, it would include the kem, the kat file, and the expected sizes.